ipam,alibabacloud: Improve event driven instance resync

jaffcheng opened the pull request:
Currently in AWS/Alibabacloud ipam modes, every time the IP allocation/release event happens, the cilium operator triggers a resync and fetches all the ENIs from instance API. In a large and frequently changing cluster, the full sync might take several tens of seconds. This slows down the IP allocation process severely and also imposes a lot of pressure on the instance API. The full sync here should be unnecessary since we only need to update the ENIs of the instance that triggered the event.

This patch introduces a new `Node.instanceSync` trigger to replace `NodeManager.resyncTrigger`. Whenever an instance is mutated due to IP pool maintenance, the trigger attempts to incrementally resynchronize the corresponding instance to the local cache. This is achieved through the newly introduced `InstanceSync` method of the `AllocationImplementation` interface.
While this feature is implemented for Alibabacloud, AWS and Azure still fall back to full resynchronization.

Here are some time cost data from an Alibabacloud cluster during different periods:

                                  full sync    InstanceSync
    time cost with ~8000  ENIs    ~20s         ~2s
    time cost with ~13000 ENIs    ~35s         ~2s

Related: #25073

```release-note
ipam,alibabacloud: Improve event driven instance resync
```

tommyp1ckles commented on pkg/ipam/node_manager.go:
```diff
@@ -370,6 +395,25 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
+				Name:        fmt.Sprintf("ipam-node-instance-sync-%s", resource.Name),
+				MinInterval: 10 * time.Millisecond,
+				TriggerFunc: func(reasons []string) {
+					if syncTime, ok := node.instanceIncrementalSync(context.TODO(), resource.InstanceID()); ok {
```
Should ensure that we constrain the timeout ourselves with the context?

tommyp1ckles commented on pkg/alibabacloud/eni/instances.go:
```diff
@@ -122,6 +123,55 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+	return true
+}
+
+func (m *InstancesManager) InstanceSync(ctx context.Context, instanceID string) time.Time {
```
opinion: using a time object to represent if the call succeeded seems a bit unintuitive to me.

jaffcheng replied:
I agree but it looks like the `Resync` interface involves many places, I'm not sure if we should change it in this PR.

tommyp1ckles commented on pkg/alibabacloud/api/api.go:
```diff
@@ -84,6 +84,27 @@ func (c *Client) GetInstances(ctx context.Context, vpcs ipamTypes.VirtualNetwork
 
+// UpdateInstance fetches the latest ENIs of the given instanceID,
+// updates the local instanceMap and return
+func (c *Client) UpdateInstance(ctx context.Context, vpcs ipamTypes.VirtualNetworkMap, subnets ipamTypes.SubnetMap, instanceID string, instances *ipamTypes.InstanceMap) (*ipamTypes.InstanceMap, error) {
```
This function seems to be very similar to GetInstances, except rather than creating a new map it populates one that's passed in. Could we simplify this somehow?

tommyp1ckles replied:
(imo: neither of these really fit into the Alibaba API interface very well which seems like it should just be an actions interface around the base alibaba API client.  Processing/population maps should maybe just be in a helper function?)

tommyp1ckles commented on pkg/alibabacloud/eni/instances.go:
```diff
@@ -122,6 +123,55 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+	}
+
+	oldInstances := m.instances.DeepCopy()
+	instances, err := m.api.UpdateInstance(ctx, vpcs, vSwitches, instanceID, oldInstances)
```
Two questions here:

* What is the implication of merging maps, could this result in leaking of old instances that may have been removed via the GetInstances call?

* Similar to my other question - this function looks almost identical to Resync, except for using UpdateInstance instead of GetInstance. I'm wondering if we could reduce the duplication a bit?

jaffcheng replied:
Thanks for the comments! I'll try to address them and refactor the code.

> What is the implication of merging maps, could this result in leaking of old instances that may have been removed via the GetInstances call?

Thanks for pointing this out. One issue I can imagine is that after this change, if an ENI on the instance is somehow detached while maintaining the IP pool, it will not be deleted from the map until the next background full resync, I think this could be fixed, will do that in the next amending.  
Other than that, I think ENIs on different nodes would not overlap and the introduced `instancesAPIResyncLock` would avoid data races.

jaffcheng committed:
ipam,alibabacloud: Improve event driven instance resync

Currently in AWS/Alibabacloud ipam modes, every time the
IP allocation/release event happens, the cilium operator
triggers a resync and fetches all the ENIs from instance
API. In a large and frequently changing cluster, the full
sync might take several tens of seconds.
This slows down the IP allocation process severely and
also imposes a lot of pressure on the instance API.
The full sync here should be unnecessary since we only
need to update the ENIs of the instance that triggered
the event.

This patch introduces a new `Node.instanceSync` trigger
to replace `NodeManager.resyncTrigger`. Whenever an
instance is mutated due to IP pool maintenance, the trigger
attempts to incrementally resynchronize the corresponding
instance to the local cache. This is achieved through the
newly introduced `InstanceSync` method of the
`AllocationImplementation` interface.
While this feature is implemented for Alibabacloud, AWS
and Azure still fall back to full resynchronization.

Here are some time cost data from an Alibabacloud cluster
during different periods:

                              full sync    InstanceSync
time cost with ~8000  ENIs    ~20s         ~2s
time cost with ~13000 ENIs    ~35s         ~2s

Related: #25073
Signed-off-by: Jaff Cheng <jaff.cheng.sh@gmail.com>

tommyp1ckles commented on pkg/alibabacloud/eni/instances.go:
```diff
@@ -99,21 +115,43 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+			"numSecurityGroups": len(securityGroups),
+		}).Info("Synchronized ENI information")
+
+		m.mutex.Lock()
```
nit: it may be better to defer the unlock here, makes inadvertently introducing lock bugs less likely.

tommyp1ckles commented on pkg/ipam/node_manager.go:
```diff
@@ -218,7 +227,17 @@ func NewNodeManager(instancesAPI AllocationImplementation, k8sAPI CiliumNodeGett
 func (n *NodeManager) instancesAPIResync(ctx context.Context) (time.Time, bool) {
+	if n.instancesAPI.SupportInstanceSync() {
+		// Full API resync should block all the instance incremental sync from other nodes
+		n.instancesAPIResyncLock.Lock()
```
I think this in this case this is fine, but this strikes me as a bit risky, theoretically (perhaps unlikely?) there's nothing to say that any future implementation of the instancesAPI couldn't alternate between returning true here in which case you could have a dangling lock.

Or to put this another way, this puts the onus on the implementer of the cloud API implementation to know the semantics of the ipam node manager - otherwise they risk introducing bugs.

tommyp1ckles commented on pkg/ipam/node_manager.go:
```diff
@@ -119,6 +119,14 @@ type AllocationImplementation interface {
+	// InstanceSync is called to sync the state of the specified instance with
+	// external APIs or systems.
+	InstanceSync(ctx context.Context, instanceID string) time.Time
+
```
Looking at this a bit more, it feels a bit cumbersome for the node manager to have to branch based on what the underlying instance API implementation does.

At the end of the day, the IPAM manager just wants to request a resync of the instance state, having a function that returns whether the implementation supports a feature feels like we're leaking implementation details a bit.  It also makes the node handler code more complex.

I wonder if this optimization couldn't be done more opaquely by the underlying implementations?

tommyp1ckles reviewed (approved):
Thanks for the changes 🥳 I think this all makes sense to me.

I'm not as familiar with the alibaba implementation, so I'm approving for Azure/general-IPAM

@christarazi do you mind taking a look as well 🙏

christarazi reviewed (approved):
LGTM thanks for the PR!

My only nit is please remove the `Fixes` from the commit msg because https://github.com/cilium/cilium/issues/25073 will not be fully resolved for all cloud-based IPAM modes which suffer from the full sync. AFAIU, this PR only resolves it for Alibaba. You can used `Related` in the commit msg instead.

gandro reviewed (approved):
Nice work, the new iteration looks much cleaner! Thanks
